File: bridge/typescript/src/core/string.ts
import type { StringId, StringPoolData } from "../_generated/core/string.js";

/** Pool for resolving stable string ids to strings. */
export class StringPool {
    readonly #textById: ReadonlyMap<StringId, string>;
    readonly #idByText: ReadonlyMap<string, StringId>;

    /** Create one string pool from exact entries. */
    constructor(strings: Iterable<readonly [StringId, string]>) {
        const textById = new Map<StringId, string>();
        const idByText = new Map<string, StringId>();

        // validate all entries while building both lookup directions
        for (const [id, text] of strings) {
            const existingText = textById.get(id);
            if (existingText !== undefined && existingText !== text) {
                throw new Error(`string id collision: ${id}`);
            }

            const existingId = idByText.get(text);
            if (existingId !== undefined && existingId !== id) {
                throw new Error(`duplicate string text with different ids: ${text}`);
            }

            textById.set(id, text);
            idByText.set(text, id);
        }

        this.#textById = textById;
        this.#idByText = idByText;
    }

    /** Create one string pool from serialized string pool data. */
    static fromData(data: StringPoolData): StringPool {
        return new StringPool(data.strings);
    }

    /** Return whether one string id is present. */
    has(id: StringId): boolean {
        return this.#textById.has(id);
    }

    /** Return the string for one string id. */
    get(id: StringId): string {
        const text = this.#textById.get(id);
        if (text === undefined) {
            throw new Error(`unknown string id: ${id}`);
        }

        return text;
    }

    /** Return the string for one string id when present. */
    getMaybe(id: StringId | undefined): string | undefined {
        return id === undefined ? undefined : this.#textById.get(id);
    }

    /** Return the stable string id for one already interned string. */
    intern(text: string): StringId {
        const id = this.#idByText.get(text);
        if (id === undefined) {
            throw new Error(`unknown string text: ${text}`);
        }

        return id;
    }

    /** Return all strings for a sequence of string ids. */
    values(ids: Iterable<StringId>): string[] {
        const values = [];

        // resolve ids in caller-provided order
        for (const id of ids) {
            values.push(this.get(id));
        }

        return values;
    }

    /** Join all strings for a sequence of string ids. */
    join(ids: Iterable<StringId>, separator = ""): string {
        return this.values(ids).join(separator);
    }

    /** Return this pool as serialized data. */
    data(): StringPoolData {
        const strings = Array.from(this.#textById.entries());

        // sort by stable string id for reproducible serialization
        strings.sort(([left], [right]) => {
            if (left < right) {
                return -1;
            } else if (left > right) {
                return 1;
            } else {
                return 0;
            }
        });

        return { strings };
    }
}

File: client/python/src/destack/core/string.py
from __future__ import annotations

from collections.abc import Iterable, Sequence
from dataclasses import dataclass

from .._generated.core.string import StringId, StringPoolData


@dataclass(frozen=True, slots=True)
class StringPool:
    """Pool for resolving stable string ids to strings."""

    _text_by_id: dict[StringId, str]
    _id_by_text: dict[str, StringId]

    def __init__(self, strings: Iterable[tuple[StringId, str]]) -> None:
        # build both lookup directions in one pass
        text_by_id: dict[StringId, str] = {}
        id_by_text: dict[str, StringId] = {}

        # validate all entries while building both lookup directions
        for id_, text in strings:
            existing_text = text_by_id.get(id_)
            if existing_text is not None and existing_text != text:
                raise ValueError(f"string id collision: {id_}")

            existing_id = id_by_text.get(text)
            if existing_id is not None and existing_id != id_:
                raise ValueError(f"duplicate string text with different ids: {text}")

            text_by_id[id_] = text
            id_by_text[text] = id_

        object.__setattr__(self, "_text_by_id", text_by_id)
        object.__setattr__(self, "_id_by_text", id_by_text)

    @classmethod
    def from_data(cls, data: StringPoolData) -> StringPool:
        """Create one string pool from serialized string pool data."""

        return cls(data.strings)

    def has(self, id_: StringId) -> bool:
        """Return whether one string id is present."""

        return id_ in self._text_by_id

    def get(self, id_: StringId) -> str:
        """Return the string for one string id."""

        text = self._text_by_id.get(id_)
        if text is None:
            raise KeyError(f"unknown string id: {id_}")

        return text

    def get_maybe(self, id_: StringId | None) -> str | None:
        """Return the string for one string id when present."""

        return None if id_ is None else self._text_by_id.get(id_)

    def intern(self, text: str) -> StringId:
        """Return the stable string id for one already interned string."""

        id_ = self._id_by_text.get(text)
        if id_ is None:
            raise KeyError(f"unknown string text: {text}")

        return id_

    def values(self, ids: Iterable[StringId]) -> list[str]:
        """Return all strings for a sequence of string ids."""

        values = []

        # resolve ids in caller-provided order
        for id_ in ids:
            values.append(self.get(id_))

        return values

    def join(self, ids: Iterable[StringId], separator: str = "") -> str:
        """Join all strings for a sequence of string ids."""

        return separator.join(self.values(ids))

    def data(self) -> StringPoolData:
        """Return this pool as serialized data."""

        strings: Sequence[tuple[StringId, str]] = sorted(self._text_by_id.items())

        return StringPoolData(strings=strings)
